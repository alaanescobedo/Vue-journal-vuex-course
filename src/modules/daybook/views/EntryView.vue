<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-light">{{ year }}</span>
      </div>

      <div>
        <input
          type="file"
          @change="onChooseImage"
          ref="imageSelector"
          v-show="false"
          accept="image/png, image/jpeg"
        />

        <button v-if="entry.id" @click="onDeleteEntry" class="btn btn-danger mx-2">
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>
        <button @click="onUploadImage" class="btn btn-primary">
          Upload Photo
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>

    <hr />

    <div class="d-flex flex-column px-3 h-75">
      <textarea
        @keydown.ctrl.enter="saveEntry"
        @keydown.ctrl.i="onUploadImage"
        v-model="entry.text"
        placeholder="Que sucedio hoy?"
      ></textarea>
    </div>

    <img
      v-if="entry.picture && !localImage"
      :src="entry.picture"
      alt="entry-picture"
      class="img-thumbnail"
    />

    <img v-if="localImage" :src="localImage" alt="entry-picture" class="img-thumbnail" />
  </template>

  <Fab icon="fa-save" @on:click="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from '@vue/runtime-core';
import { mapActions, mapGetters } from 'vuex';
import getDayMonthYear from '../helpers/getDayMonthYear';
import Swal from 'sweetalert2';
import uploadImage from '../helpers/uploadImage';

export default {
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      entry: null,
      localImage: null,
      fileImage: null,
    };
  },
  components: {
    Fab: defineAsyncComponent(() => import('../components/Fab.vue')),
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    year() {
      const { year } = getDayMonthYear(this.entry.date);
      return year;
    },
  },
  methods: {
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
    loadEntry() {
      let entry;
      if (this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryById(this.id);
        this.localImage = null;
        if (!entry) return this.$router.push({ name: 'no-entry' });
      }
      this.entry = entry;
    },
    async saveEntry() {
      new Swal({
        title: 'Espere por favor',
        allowOutsideClick: false,
      });
      Swal.showLoading();

      const picture = await uploadImage(this.fileImage);
      this.entry.picture = picture;

      if (this.entry.id) {
        //Actualizar entrada
        await this.updateEntry(this.entry);
        Swal.fire('Guardado', 'Entrada actualizada con exito', 'success');
      } else {
        //Crear nueva entrada
        const id = await this.createEntry(this.entry);
        this.$router.push({ name: 'entry', params: { id } });
        Swal.fire('Guardado', 'Entrada registrada con exito', 'success');
      }

      this.fileImage = null;
    },
    async onDeleteEntry() {
      const { isConfirmed } = await Swal.fire({
        title: 'Estas seguro?',
        text: 'Una vez borrado, no se puede recuperar',
        showDenyButton: true,
        confirmButtonText: 'Si, estoy seguro',
      });

      if (isConfirmed) {
        new Swal({
          title: 'Espere por favor',
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({ name: 'no-entry' });
        Swal.fire('Eliminado', '', 'success');
      }
    },
    onChooseImage(event) {
      const fileImage = event.target.files[0];

      if (!fileImage) {
        this.localImage = null;
        this.fileImage = null;
        return;
      }

      this.fileImage = fileImage;

      const fr = new FileReader();
      fr.onload = () => (this.localImage = fr.result);
      fr.readAsDataURL(fileImage);
    },
    onUploadImage() {
      this.$refs.imageSelector.click();
    },
  },
  created() {
    this.loadEntry();
  },
  watch: {
    id() {
      this.loadEntry();
    },
  },
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: #000, $alpha: 0.2);
}
</style>
