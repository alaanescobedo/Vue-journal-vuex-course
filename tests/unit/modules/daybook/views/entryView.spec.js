import EntryView from '@/modules/daybook/views/EntryView';
import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import journalModule from '@/modules/daybook/store/journal';
import { journalState } from '../../../mock/test-journal-state';
import Swal from 'sweetalert2';

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journalModule,
        state: { ...initialState },
      },
    },
  });

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));

describe('Entry View', () => {
  const store = createVuexStore(journalState);
  store.dispatch = jest.fn();
  const mockRouter = {
    push: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryView, {
      props: {
        id: '-Mi023nECy35q-nhyISU',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
  });

  test('debe redireccionar al usuario si el Id no existe', () => {
    shallowMount(EntryView, {
      props: {
        id: 'Este id no existe',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' });
  });

  test('debe mostrar la entrada correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });

  test('debe de borrar la entrada y salir', async () => {
    Swal.fire.mockReturnValue(Promise.resolve({ isConfirmed: true }));

    const buttonDelete = wrapper.find('[data-testId="btn-delete"]');
    await buttonDelete.trigger('click');

    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Estas seguro?',
      text: 'Una vez borrado, no se puede recuperar',
      showDenyButton: true,
      confirmButtonText: 'Si, estoy seguro',
    });

    expect(mockRouter.push).toHaveBeenCalled();
    expect(mockRouter.push).toHaveBeenCalledTimes(1);

    expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-Mi023nECy35q-nhyISU');
  });
});
