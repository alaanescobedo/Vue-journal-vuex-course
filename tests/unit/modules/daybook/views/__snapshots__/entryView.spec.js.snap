// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry View debe mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">25</span><span class="mx-1 fs-3">Agosto</span><span class="mx-2 fs-4 fw-light">2021, Mi√©rcoles</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button data-testid="btn-delete" class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload Photo <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Que sucedio hoy?"></textarea></div><img src="https://res.cloudinary.com/desxpqy6n/image/upload/v1629919238/Test-1/ui2zibpqlaxujasrl2vx.png" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
